packages/flutter_tools/gradle/src/main/kotlin/dependency_version_checker.gradle.kts

```diff
@@ -0,0 +1,270 @@
+                        "version checking.")
+            }
+        }
+
+        // https://docs.gradle.org/current/kotlin-dsl/gradle/org.gradle.api.invocation/-gradle/index.html#-837060600%2FFunctions%2F-1793262594
+        fun getGradleVersion(project : Project) : Version {
+            return Version.fromString(project.gradle.getGradleVersion())
+        }
+
+        // https://docs.gradle.org/current/kotlin-dsl/gradle/org.gradle.api/-java-version/index.html#-1790786897%2FFunctions%2F-1793262594
+        fun getJavaVersion(project : Project) : JavaVersion {
+            return JavaVersion.current()
+        }
+
+        // This approach is taken from AGP's own version checking plugin:
+        // https://android.googlesource.com/platform/tools/base/+/1839aa23b8dc562005e2f0f0cc8e8b4c5caa37d0/build-system/gradle-core/src/main/java/com/android/build/gradle/internal/utils/agpVersionChecker.kt#58.
+        fun getAGPVersion(project: Project): Version? {
+            var agpVersion: Version? = null
+            try {
+                agpVersion = Version.fromString(
+                    project.plugins.getPlugin("com.android.base")::class.java.classLoader.loadClass(
+                        com.android.Version::class.java.name
+                    ).fields.find { it.name == "ANDROID_GRADLE_PLUGIN_VERSION" }!!
+                        .get(null) as String
+                )
+            } catch (ignored: ClassNotFoundException) {
+                // Use deprecated Version class as it exists in older AGP (com.android.Version) does
+                // not exist in those versions.
+                agpVersion = Version.fromString(
+                    project.plugins.getPlugin("com.android.base")::class.java.classLoader.loadClass(
+                        com.android.builder.model.Version::class.java.name
+                    ).fields.find { it.name == "ANDROID_GRADLE_PLUGIN_VERSION" }!!
+                        .get(null) as String
+                )
+            }
+            return agpVersion
+        }
+
+        fun getKGPVersion(project : Project) : Version? {
+            // This property corresponds to application of the Kotlin Gradle plugin in the
+            // top-level build.gradle file.
+            if (project.hasProperty("kotlin_version")) {
+                return Version.fromString(project.properties.get("kotlin_version") as String)
+            }
+            val kotlinPlugin = project.getPlugins()
+                .findPlugin(KotlinAndroidPluginWrapper::class.java)
+            val versionfield =
+                kotlinPlugin?.javaClass?.kotlin?.members?.first { it.name == "pluginVersion" || it.name == "kotlinPluginVersion" }
+            val versionString = versionfield?.call(kotlinPlugin)
+            if (versionString == null) {
+                return null
+            } else {
+                return Version.fromString(versionfield!!.call(kotlinPlugin) as String)
```

Comment:
Why cant you use `versionString` here instead of using `versionfield!!.call(kotlinPlugin)`?

Reply:
I can (and should), thanks for pointing it out! changed